From 62bb1f39dd9ff1652117c2d88444eb70438c5fff Mon Sep 17 00:00:00 2001
Message-ID: <62bb1f39dd9ff1652117c2d88444eb70438c5fff.1779975966.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Fri, 22 May 2026 11:23:17 +0200
Subject: [PATCH 12/30] virtio-msg-bus: Add must-deliver sends

Valid virtio-msg request handling depends on sending the matching response,
and generated events also need reliable delivery. Best-effort transport send
behavior is not enough for that traffic.

Add a separate bus callback for messages that must eventually be delivered
and route required responses and events through it. Transports that do not
need special retry handling continue to use the normal send path.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio-msg-bus.c         | 18 +++++++++++++
 hw/virtio/virtio-msg.c             | 41 +++++++++++++++++++-----------
 include/hw/virtio/virtio-msg-bus.h |  7 +++++
 3 files changed, 51 insertions(+), 15 deletions(-)

diff --git a/hw/virtio/virtio-msg-bus.c b/hw/virtio/virtio-msg-bus.c
index abf2218d9b21..4e968eb0323d 100644
--- a/hw/virtio/virtio-msg-bus.c
+++ b/hw/virtio/virtio-msg-bus.c
@@ -53,6 +53,24 @@ int virtio_msg_bus_send(BusState *bus, VirtIOMSG *msg_req)
     return r;
 }
 
+int virtio_msg_bus_send_must_deliver(BusState *bus, VirtIOMSG *msg_req)
+{
+    VirtIOMSGBusDevice *bd = virtio_msg_bus_get_device(bus);
+    VirtIOMSGBusDeviceClass *bdc;
+
+    if (!virtio_msg_bus_peer_ready(bd)) {
+        return VIRTIO_MSG_ERROR_RETRY;
+    }
+
+    bdc = VIRTIO_MSG_BUS_DEVICE_CLASS(object_get_class(OBJECT(bd)));
+
+    if (bdc->send_must_deliver) {
+        return bdc->send_must_deliver(bd, msg_req);
+    }
+
+    return virtio_msg_bus_send(bus, msg_req);
+}
+
 static void virtio_msg_bus_class_init(ObjectClass *klass, const void *data)
 {
     BusClass *bc = BUS_CLASS(klass);
diff --git a/hw/virtio/virtio-msg.c b/hw/virtio/virtio-msg.c
index e77bd47e0f74..5c21780364f1 100644
--- a/hw/virtio/virtio-msg.c
+++ b/hw/virtio/virtio-msg.c
@@ -22,6 +22,17 @@
 
 #define VIRTIO_MSG_VENDOR_ID 0x554D4551 /* 'QEMU' */
 
+static void virtio_msg_send_must_deliver(VirtIOMSGProxy *s, VirtIOMSG *msg)
+{
+    int ret = virtio_msg_bus_send_must_deliver(&s->msg_bus, msg);
+
+    if (ret != VIRTIO_MSG_NO_ERROR) {
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: failed to send required message id %u: ret=%d\n",
+                      __func__, msg->msg_id, ret);
+    }
+}
+
 static bool virtio_msg_bad(VirtIOMSGProxy *s, VirtIOMSG *msg)
 {
     bool drop = false;
@@ -137,7 +148,7 @@ static void virtio_msg_bus_get_devices(VirtIOMSGProxy *s,
     virtio_msg_pack_bus_get_devices_resp(&msg_resp,
                                          offset, num, next_offset,
                                          data);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_bus_ping(VirtIOMSGProxy *s, VirtIOMSG *msg)
@@ -145,7 +156,7 @@ static void virtio_msg_bus_ping(VirtIOMSGProxy *s, VirtIOMSG *msg)
     VirtIOMSG msg_resp;
 
     virtio_msg_pack_bus_ping_resp(&msg_resp, msg->bus_ping.data);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_device_info(VirtIOMSGProxy *s,
@@ -174,7 +185,7 @@ static void virtio_msg_device_info(VirtIOMSGProxy *s,
                                          config_len,
                                          max_vqs,
                                          0, 0);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_get_features(VirtIOMSGProxy *s,
@@ -203,7 +214,7 @@ static void virtio_msg_get_features(VirtIOMSGProxy *s,
     /* If index is out of bounds, we respond with num=0, f=0.  */
     virtio_msg_pack_get_features_resp(&msg_resp, msg->dev_num, msg->token,
                                       index, num, f);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_set_features(VirtIOMSGProxy *s,
@@ -229,7 +240,7 @@ static void virtio_msg_set_features(VirtIOMSGProxy *s,
     s->devs[msg->dev_num].guest_features = f;
 
     virtio_msg_pack_set_features_resp(&msg_resp, msg->dev_num, msg->token);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_soft_reset(VirtIOMSGProxy *s, uint16_t dev_num)
@@ -272,7 +283,7 @@ static void virtio_msg_set_device_status(VirtIOMSGProxy *s,
 
     virtio_msg_pack_set_device_status_resp(&msg_resp, msg->dev_num, msg->token,
                                            vdev->status);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_get_device_status(VirtIOMSGProxy *s,
@@ -284,7 +295,7 @@ static void virtio_msg_get_device_status(VirtIOMSGProxy *s,
 
     virtio_msg_pack_get_device_status_resp(&msg_resp, msg->dev_num, msg->token,
                                            status);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_get_config(VirtIOMSGProxy *s,
@@ -314,7 +325,7 @@ static void virtio_msg_get_config(VirtIOMSGProxy *s,
     virtio_msg_pack_get_config_resp(&msg_resp, msg->dev_num, msg->token,
                                     size, offset,
                                     generation, data);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_set_config(VirtIOMSGProxy *s,
@@ -338,7 +349,7 @@ static void virtio_msg_set_config(VirtIOMSGProxy *s,
     virtio_msg_pack_set_config_resp(&msg_resp, msg->dev_num, msg->token,
                                     size, offset,
                                     generation, data);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_get_vqueue(VirtIOMSGProxy *s,
@@ -366,7 +377,7 @@ static void virtio_msg_get_vqueue(VirtIOMSGProxy *s,
                                         index, 0, 0, 0, 0, 0);
     }
 
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_set_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
@@ -389,7 +400,7 @@ static void virtio_msg_set_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
     virtio_queue_enable(vdev, index);
 
     virtio_msg_pack_set_vqueue_resp(&msg_resp, msg->dev_num, msg->token);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_reset_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
@@ -404,7 +415,7 @@ static void virtio_msg_reset_vqueue(VirtIOMSGProxy *s, VirtIOMSG *msg)
     virtio_queue_reset(vdev, msg->reset_vqueue.index);
 
     virtio_msg_pack_reset_vqueue_resp(&msg_resp, msg->dev_num, msg->token);
-    virtio_msg_bus_send(&s->msg_bus, &msg_resp);
+    virtio_msg_send_must_deliver(s, &msg_resp);
 }
 
 static void virtio_msg_event_avail(VirtIOMSGProxy *s,
@@ -425,7 +436,7 @@ static void virtio_msg_event_avail(VirtIOMSGProxy *s,
         virtio_msg_pack_event_config(&msg_ev, msg->dev_num,
                                      vdev->status, vdev->generation,
                                      0, 0, NULL);
-        virtio_msg_bus_send(&s->msg_bus, &msg_ev);
+        virtio_msg_send_must_deliver(s, &msg_ev);
         return;
     }
 
@@ -533,7 +544,7 @@ static void virtio_msg_notify(DeviceState *opaque, uint16_t vector)
 
     if (vector < VIRTIO_QUEUE_MAX) {
         virtio_msg_pack_event_used(&msg, mdev->dev_num, vector);
-        virtio_msg_bus_send(&s->msg_bus, &msg);
+        virtio_msg_send_must_deliver(s, &msg);
         return;
     }
 
@@ -541,7 +552,7 @@ static void virtio_msg_notify(DeviceState *opaque, uint16_t vector)
         virtio_msg_pack_event_config(&msg, mdev->dev_num,
                                      vdev->status, vdev->generation,
                                      0, 0, NULL);
-        virtio_msg_bus_send(&s->msg_bus, &msg);
+        virtio_msg_send_must_deliver(s, &msg);
         return;
     }
 }
diff --git a/include/hw/virtio/virtio-msg-bus.h b/include/hw/virtio/virtio-msg-bus.h
index d99d5369aea9..d85062d36c9f 100644
--- a/include/hw/virtio/virtio-msg-bus.h
+++ b/include/hw/virtio/virtio-msg-bus.h
@@ -49,6 +49,12 @@ struct VirtIOMSGBusDeviceClass {
      * Called by the transport to send a message.
      */
     int (*send)(VirtIOMSGBusDevice *bd, VirtIOMSG *msg_req);
+
+    /*
+     * Called by the transport to send required traffic that must be
+     * delivered eventually.
+     */
+    int (*send_must_deliver)(VirtIOMSGBusDevice *bd, VirtIOMSG *msg_req);
 };
 
 typedef struct VirtIOMSGBusDevice {
@@ -96,6 +102,7 @@ virtio_msg_bus_receive(VirtIOMSGBusDevice *bd, VirtIOMSG *msg)
 }
 
 int virtio_msg_bus_send(BusState *bus, VirtIOMSG *msg_req);
+int virtio_msg_bus_send_must_deliver(BusState *bus, VirtIOMSG *msg_req);
 
 static inline AddressSpace *virtio_msg_bus_get_remote_as(BusState *bus)
 {
-- 
2.53.0

